private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,665 @@
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule"
+	"github.com/bufbuild/buf/private/gen/data/datawkt"
+	"github.com/bufbuild/buf/private/pkg/app/appext"
+	"github.com/bufbuild/buf/private/pkg/ioext"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/syserror"
+	"github.com/bufbuild/buf/private/pkg/tracing"
+	"github.com/fsnotify/fsnotify"
+	"go.lsp.dev/jsonrpc2"
+	"go.lsp.dev/protocol"
+	"go.uber.org/zap"
+)
+
+var (
+	// wellKnownTypesDescriptorProtoPath is the path of the descriptor.proto well-known types file.
+	wellKnownTypesDescriptorProtoPath = normalpath.Join("google", "protobuf", "descriptor.proto")
+	// lspWellKnownTypesCacheRelDirPath is the relative path to the cache directory for materialized
+	// well-known types .proto files.
+	lspWellKnownTypesCacheRelDirPath = normalpath.Join("lsp", "wkt")
+	// v3CacheModuleRelDirPath is the relative path to the cache directory in its newest iteration.
+	// NOTE: This needs to be kept in sync with module_data_provider.
+	v3CacheModuleRelDirPath = normalpath.Join("v3", "modules")
+	// v3CacheExternalModuleDataFilesDir is the subdirectory within a module commit's cache
+	// directory where its external data (e.g. proto files) is stored.
+	v3CacheExternalModuleDataFilesDir = "files"
+)
+
+// NewServer returns a new LSP server for the jsonrpc connection.
+func NewServer(
+	ctx context.Context,
+	jsonrpc2Conn jsonrpc2.Conn,
+	container appext.Container,
+	controller bufctl.Controller,
+) (protocol.Server, error) {
+	return newServer(ctx, jsonrpc2Conn, container, controller)
+}
+
+// *** PRIVATE ***
+
+type server struct {
+	nopServer
+
+	jsonrpc2Conn            jsonrpc2.Conn
+	logger                  *zap.Logger
+	container               appext.Container
+	controller              bufctl.Controller
+	lintHandler             buflint.Handler
+	breakingHandler         bufbreaking.Handler
+	fileCache               map[string]*fileEntry
+	fileWatcher             *fsnotify.Watcher
+	wellKnownTypesModuleSet bufmodule.ModuleSet
+	wellKnownTypesResolver  moduleSetResolver
+
+	folders   []protocol.WorkspaceFolder
+	clientCap protocol.ClientCapabilities
+
+	lock sync.Mutex
+}
+
+func newServer(
+	ctx context.Context,
+	jsonrpc2Conn jsonrpc2.Conn,
+	container appext.Container,
+	controller bufctl.Controller,
+) (*server, error) {
+	logger := container.Logger()
+	tracer := tracing.NewTracer(container.Tracer())
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	wellKnownTypesResolver := newModuleSetResolver(func() (bufmodule.ModuleSet, error) {
+		moduleSetBuilder := bufmodule.NewModuleSetBuilder(ctx, logger, bufmodule.NopModuleDataProvider)
+		moduleSetBuilder.AddLocalModule(
+			datawkt.ReadBucket,
+			".",
+			true,
+		)
+		return moduleSetBuilder.Build()
+	})
+	wellKnownTypesModuleSet, err := wellKnownTypesResolver.ModuleSet()
+	if err != nil {
+		return nil, err
+	}
+	buflsp := &server{
```

Comment:
This confused me as the variable name is shadowing the package name - why not just name it server?